### PR TITLE
Bug/19702 bbq.js proto secure

### PR DESCRIPTION
**Pull request by kevin-foster-uk**
This PR fixes a security issue reported to LimeSurvey (https://bugs.limesurvey.org/view.php?id=19702).

We have patched jquery-bbq.js to prevent injection of __proto__. 

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ❌ (tests are failing but its not related to this PR)

### CODE REVIEW

**Review by Arhell — COMMENTED** [review]
@marcovtwout

File: framework/web/js/source/jquery.ba-bbq.js [hunk 1]
```diff
@@ -480,6 +482,10 @@
         // into its component parts.
         keys = key.split( '][' ),
         keys_last = keys.length - 1;
+
+      if (prohibitedKeys.includes(key)) {
+        return;
+      }
```
Comment (marcovtwout) [inline framework/web/js/source/jquery.ba-bbq.js:488]: Indentation not correct.

File: framework/web/js/source/jquery.ba-bbq.js [hunk 2]
```diff
@@ -1282,96 +1293,6 @@
       timeout_id = setTimeout( poll, $.fn[ str_hashchange ].delay );
     };
     
-    // vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
-    // vvvvvvvvvvvvvvvvvvv REMOVE IF NOT SUPPORTING IE6/7/8 vvvvvvvvvvvvvvvvvvv
-    // vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
```
Comment (marcovtwout) [inline framework/web/js/source/jquery.ba-bbq.js:1287]: This removal does not look relevant to the security fix.

File: framework/web/js/source/jquery.ba-bbq.min.js [hunk 3]
```diff
@@ -1,18 +1 @@
...
...doc_mode=doc.documentMode,supports_onhashchange="on"+str_hashchange in window&&(doc_mode===undefined||doc_mode>7);function get_fragment(url){url=url||location.href;return"#"+url.replace(/^[^#]*#?(.*)$/,"$1")}$.fn[str_hashchange]=function(fn){return fn?this.bind(str_hashchange,fn):this.trigger(str_hashchange)};$.fn[str_hashchange].delay=50;special[str_hashchange]=$.extend(special[str_hashchange],{setup:function(){if(supports_onhashchange){return false}$(fake_onhashchange.start)},teardown:function(){if(supports_onhashchange){return false}$(fake_onhashchange.stop)}});fake_onhashchange=function(){var self={},timeout_id,last_hash=get_fragment(),fn_retval=function(val){return val},history_set=fn_retval,history_get=fn_retval;self.start=function(){timeout_id||poll()};self.stop=function(){timeout_id&&clearTimeout(timeout_id);timeout_id=undefined};function poll(){var hash=get_fragment(),history_hash=history_get(last_hash);if(hash!==last_hash){history_set(last_hash=hash,history_hash);$(window).trigger(str_hashchange)}else if(history_hash!==last_hash){location.href=location.href.replace(/#.*/,"")+history_hash}timeout_id=setTimeout(poll,$.fn[str_hashchange].delay)}return self}()})(jQuery,this);
```
Comment (marcovtwout) [inline framework/web/js/source/jquery.ba-bbq.min.js:1]: This minification looks totally different from the original, and also the license headers are missing. Is it possible to use the same minification method?

File: framework/web/js/source/jquery.ba-bbq.js (hunk 2 above)
Comment (rob006) [reply framework/web/js/source/jquery.ba-bbq.js:1287]: I would remove it anyway for the sake of removing amount of code that may need to be fixed/tested for this issue. IMO supporting IE does not make sense anymore in 2024.

File: framework/web/js/source/jquery.ba-bbq.js (hunk 1 above)
Comment (kevin-foster-uk) [reply framework/web/js/source/jquery.ba-bbq.js]: I have fixed the formatting. The indentation still looks a bit weird but it is correct if you look at all of the context above.

File: framework/web/js/source/jquery.ba-bbq.min.js (hunk 3 above)
Comment (kevin-foster-uk) [reply framework/web/js/source/jquery.ba-bbq.min.js]: I cant find any information about how the original was minified. The project does not contain a script for this and also didn't include any documentation about this on its GitHub or website.

I created an issue on the source project asking for details:
https://github.com/cowboy/jquery-bbq/issues/64

I have used uglify-js:
`uglifyjs jquery.ba-bbq.js --compress -o jquery.ba-bbq.min.js`

File: framework/web/js/source/jquery.ba-bbq.js (hunk 2 above)
Comment (marcovtwout) [reply framework/web/js/source/jquery.ba-bbq.js:1287]: Agreed. I see this was also suggested upstream: https://github.com/cowboy/jquery-bbq/pull/48/files

File: framework/web/js/source/jquery.ba-bbq.min.js (hunk 3 above)
Comment (marcovtwout) [reply framework/web/js/source/jquery.ba-bbq.min.js]: Confirmed that the new uglifiy output matches the source.

**Comment by kevin-foster-uk** [comment]
> The LimeSurvey link is not available without logging in. Is there a public version?
> 
> The source change looks good but there are some points to address.

Login is required but it is open for anyone to create an account. You just need to provide your name and email address.

**Comment by marcovtwout** [comment]
I checked the upstream project where this issue was also found, but never fixed fully. For the sake of any other projects using jquery-bbq, I suggest to share your solution upstream also.

Links:
- CVE: https://www.cve.org/CVERecord?id=CVE-2021-20086
- Upstream issue: https://github.com/cowboy/jquery-bbq/issues/62
- Upstream PR with partial fix: https://github.com/cowboy/jquery-bbq/pull/61

Other noteworthy info:
- Latest upstream version: 1.3pre, Last updated: 8/26/2010
- Last version in the wild: 1.4pre, Last updated: 1/15/2013. Removed $.browser reference to work with jQuery 1.9+. Yii is currently using this version although the comment in the minified file wrongly states it's 1.3pre.

File: framework/web/js/source/jquery.ba-bbq.min.js [hunk 4]
```diff
@@ -1,18 +1,17 @@
 /*
  * jQuery BBQ: Back Button & Query Library - v1.3pre - 8/26/2010
- * https://benalman.com/projects/jquery-bbq-plugin/
+ * http://benalman.com/projects/jquery-bbq-plugin/
```
Comment (marcovtwout) [inline framework/web/js/source/jquery.ba-bbq.min.js:3]: Could you update `v1.3pre` to `v1.4pre`? (was already the case but documented wrongly)
And please revert the 4 lines that change `https` to `http`

File: framework/web/js/source/jquery.ba-bbq.js [hunk 5]
```diff
@@ -465,7 +465,9 @@
   //  (Object) An object representing the deserialized params string.
   
   $.deparam = jq_deparam = function( params, coerce ) {
```
Comment (marcovtwout) [inline framework/web/js/source/jquery.ba-bbq.js:467]: Perhaps you can add a comment header block in the first lines of this file, stating the changes made compared to 1.4pre. Something along the lines of:

```
Additional changes to this file:

- Removed support for IE8 and below
- Fixed prototype pollution (CVE-2021-20086)
```

**Review by marcovtwout — CHANGES_REQUESTED** [review]

File: framework/web/js/source/jquery.ba-bbq.min.js [hunk 6]
```diff
@@ -1,18 +1,17 @@
 /*
```
Comment (marcovtwout) [inline framework/web/js/source/jquery.ba-bbq.min.js:1]: Perhaps add an addition comment that says:

```
Manually re-uglified with `uglifyjs jquery.ba-bbq.js --compress -o jquery.ba-bbq.min.js`
```

Comment (rob006) [reply framework/web/js/source/jquery.ba-bbq.min.js:1]: I don't think that such information should be in minified version. If we want to document how to minify it it the future, it should be rather in source file.

**Comment by marcovtwout** [comment]
@kevin-foster-uk could you look at the above? When this PR is done I want to prepare a new yii release.

**Comment by kevin-foster-uk** [comment]
> @kevin-foster-uk could you look at the above? When this PR is done I can prepare a new yii release.

Sorry for the delay I have been on vacation. Will update asap.

**Review by marcovtwout — APPROVED** [review]

**Comment by marcovtwout** [comment]
Thanks! I'll update the CHANGELOG on master.

**Pull request merged**